A prototype for revised suite and task questions, redux (#514)
GEOS-ESM/swell: docs/adding_a_suite.md

Task question lists are stored in `src/swell/tasks/task_questions.py`

`QuestionList` objects store and handle questions in an object-oriented manner. They can store questions directly, or store other lists to use their questions. Here is an example of a question list for a task: 

```python
    BuildJediByLinking = QuestionList(
        list_name="BuildJediByLinking",
        questions=[
            qd.existing_jedi_build_directory(),
            qd.existing_jedi_build_directory_pinned(),
            qd.jedi_build_method()
        ]
    )
```

During experiment creation, Swell scans the suite's `flow.cylc` file to find all of the tasks used in the workflow. It then finds the corresponding task lists in `src/swell/tasks/task_questions.py`, and fits together a list of uniquely named questions from all of the lists. Questions have a priority depending on order. In the case of duplicate questions, those further DOWN the list take priority. For this reason, it is NOT RECOMMENDED to set different default values for tasks in `task_questions.py`, since questions may be overridden by a questions in a different task. 

> In this question infrastructure, suites priority over tasks. Any question specified in a suite configuration will override the default value for a question in one of its member tasks. This allows for easily setting different configurations for suites without having to specify redundant questions. For ease of use, model-dependent questions can be assigned directly in their respective lists.

Reviewer: ```suggestion
In this question infrastructure, **suites take priority over tasks**. Any question specified in a suite configuration will override the default value for a question in one of its member tasks. This allows for easily setting different configurations for suites without having to specify redundant questions. For ease of use, model-dependent questions can be assigned directly in their respective lists.
```

Consider the following example of suite questions for `3dvar` (in python, variable names cannot begin with digits):

```python

> class SuiteQuestions(QuestionContainer, Enum):

Reviewer: For clarity:

```suggestion
from swell.utilities.question_defaults import QuestionDefaults as qd
from swell.suites.suite_questions import SuiteQuestions as sq

class SuiteQuestions(QuestionContainer, Enum):
```

Reviewer: Also, why do we inherit from `Enum` here? (Not saying we shouldn't --- just curious to know the reason).

Author: > Also, why do we inherit from `Enum` here? (Not saying we shouldn't --- just curious to know the reason).

I think it started as me slightly misunderstanding what we should be using enums for, but there are some minor syntactical conveniences offered by enum. __getitem__  lets you get attributes similar to a dictionary, rather than using getattr (though you need to use .value to get the actual object), the _member_names_ attributes is also quite useful in this case

```
[scheduling]

    initial cycle point = 2021-07-01T12:00:00Z
    final cycle point = 2021-07-01T12:00:00Z
    runahead limit = P4

    [[graph]]
        R1 = """
            # Triggers for non cycle time dependent tasks
            # -------------------------------------------
            # Clone JEDI source code
            CloneJedi

            # Build JEDI source code by linking
            CloneJedi => BuildJediByLinking?

            # If not able to link to build create the build
            BuildJediByLinking:fail? => BuildJedi

        """
```
For context, tasks in `R1` are run as setup for later tasks. Here, `CloneJedi` is triggered first. When `CloneJedi` completes, `BuildJediByLinking` is triggered, but the question mark denotes that it is optional to complete. Should it fail, the task `BuildJedi` is triggered.

The actions associated with these tasks is defined in the `runtime` section of `flow.cylc`. Here is the runtime section for the tasks above:

```
[[runtime]]

...

> [runtime]

Reviewer: Two `[runtime]` sections. Is this a typo?

Reviewer: Or rather, I suspect the `[[runtime]]` above is a typo.

- widget_type: A custom enum specifying the data type of the answer, as well as the way the question will be asked on the command line interface. Options include drop lists, check lists, and direct entry.
- prompt: A sentence or two describing the question.
- question_type: "task" or "suite", use the `SuiteQuestion` or `TaskQuestion` parent classes to set this automatically.
- depends: A dictionary specifying default values to be used if certain other questions are asked.
- models: List of models the question applies to, can also use `["all_models"]` or `None`.
- ask_question: (Currently unused)
- options: List of valid options for the answer

An example question class:
```python
    @dataclass
    class existing_jedi_build_directory(TaskQuestion):
        default_value: str = "defer_to_platform"
        question_name: str = "existing_jedi_build_directory"
        ask_question: bool = True
        depends: Dict = field(default_factory=lambda: {
            "jedi_build_method": "use_existing"
        })
        prompt: str = "What is the path to the existing JEDI build directory?"
        widget_type: WidgetType = WidgetType.STRING
```

> Calling `existing_jedi_build_directory()` creates an object matching the default values for the class. Fields set as `defer_to_platform`, `defer_to_model`, and `defer_to_code` are later override with corresponding values sourced from:

Reviewer: ```suggestion
Calling `existing_jedi_build_directory()` creates an object matching the default values for the class. The specific values `defer_to_platform`, `defer_to_model`, and `defer_to_code` are special values that will be substituted as follows, depending on the `platform` and `model`: 
```

Reviewer: Not for this PR, but at some point, it would be nice to think about whether this could be done more elegantly, e.g., via a `field` constructor along the lines of:

```python
  default_value: str = field(default_factory = lambda: f"src/swell/deployment/platforms/{platform}")
```

That presupposes a way to access a variable called `platform` at runtime. One way to do this is to add `platform` as a required field and then use a [`__post_init__` hook](https://docs.python.org/3/library/dataclasses.html#post-init-processing) to fix these values. But that's really annoying if we have to pass platform explicitly into every one of these. Another possibility is to get clever with the `field(default_factory =. ...)` such that it is smart about detecting the platform...but that effectively introduces a (small) side effect, which may not be ideal either. Something to think about!

Author: Agreed. Fortunately `platform` is specified at swell main and should be accessible at the point of experiment construction. I'm not super familiar with what `field` constructors are capable of, but it'd be great to be able to grab values easily like that

# A Guide to Suite Configurations in Swell

# Introduction

This document provides a high-level overview of the structure of a Swell suite, with the goal of providing context and basic instructions to a user looking to add a suite to Swell. It can also serve as a primer for users less familiar to Swell. 

> ## Definitions

Reviewer 2: While I really appreciate the long descriptions here, I suggest we have these nuggets of information in a more conspicuous location. There are some reduntant parts in `getting started`part of the documentation and these bits and pieces here could replace those sections.

I'm not suggesting these changes for this PR, it could be part of the demonstration preparation while we update the documentation.

Reviewer 2: I'm not saying just the `Definitions` there is real useful information all over the `adding_a_suite.md` bit

Author: Yeah, I'll rethink the structure of this a bit when I take a look at the docs

    models: Optional[list] = None
    ask_question: bool = False
    options: Optional[str] = None
```
Arguments:
- default_value: default value for the answer
- question_name: name of the question (should usually match the class).
- widget_type: A custom enum specifying the data type of the answer, as well as the way the question will be asked on the command line interface. Options include drop lists, check lists, and direct entry.
- prompt: A sentence or two describing the question.
- question_type: "task" or "suite", use the `SuiteQuestion` or `TaskQuestion` parent classes to set this automatically.
- depends: A dictionary specifying default values to be used if certain other questions are asked.
- models: List of models the question applies to, can also use `["all_models"]` or `None`.
- ask_question: (Currently unused)
- options: List of valid options for the answer

An example question class:
```python
    @dataclass
    class existing_jedi_build_directory(TaskQuestion):
        default_value: str = "defer_to_platform"
        question_name: str = "existing_jedi_build_directory"
        ask_question: bool = True
        depends: Dict = field(default_factory=lambda: {

>             "jedi_build_method": "use_existing"

Reviewer 2: Just a minor comment, since other developers might would have to use this, could we clarify the purpose of using lambda here somewhere?

Reviewer: ```suggestion
        # Need to do a lambda here because dataclass fields cannot be
        # initialized to mutable types (like dict or list).
        # https://docs.python.org/3/library/dataclasses.html#mutable-default-values
        depends: Dict = field(default_factory=lambda: {
            "jedi_build_method": "use_existing"
        })
```

In this question infrastructure, **suites take priority over tasks**. Any question specified in a suite configuration will override the default value for a question in one of its member tasks. This allows for easily setting different configurations for suites without having to specify redundant questions. For ease of use, model-dependent questions can be assigned directly in their respective lists.

Consider the following example of suite questions for `3dvar` (in python, variable names cannot begin with digits):

```python
from swell.utilities.question_defaults import QuestionDefaults as qd
from swell.suites.suite_questions import SuiteQuestions as sq

> 
> from swell.utilities.question_defaults import QuestionDefaults as qd
> from swell.suites.suite_questions import SuiteQuestions as sq

Reviewer 2: ```suggestion
```
repetition
